lib/puppet/functions/group_by.rb

```diff
@@ -0,0 +1,19 @@
+# frozen_string_literal: true
+
+# Groups the collection by result of the block. Returns a hash where the keys are the evaluated result from the block
+# and the values are arrays of elements in the collection that correspond to the key.
+Puppet::Functions.create_function(:group_by) do
+  # @param collection A collection of things to group.
+  # @example Group strings by length, results in e.g. { 1 => [a, b], 2 => [ab] }
+  #   [a, b, ab].group_by |$s| { $s.length }
+  dispatch :group_by do
+    required_param 'Iterable', :collection
+    block_param
+    return_type 'Hash'
+  end
+
+  def group_by(collection, &block)
+    collection.group_by(&block).freeze
```

Comment:
This will not work for all data types that are also Iterable (String for example).